### PR TITLE
Review custom object system name and label

### PR DESCRIPTION
**Pull request by cconard96**
## Checklist before requesting a review

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [x] I have added tests that prove my fix is effective or that my feature works.

## Description

Adjust how Custom Assets and Custom Dropdowns have their system name / default label handled to match closer to what was done with Custom Fields.

- [x] Add a "Label" field which is shown before the system name field and is the primary input for users
- [x] The "system name" field is automatically calculated from the label field unless the user overrides it by typing in the system name field manually
- [x] Use the "Label" as the fallback translation
- [x] Keep the tooltip helper for the system name, but keep it simple
- [x] Have a proper transliteration, accented characters should transform to their non-accented version (ex é -> e)
- [x] Rename the `name` column for Custom Field definitions to `system_name`

### CODE REVIEW

**Comment by trasher** [comment]
Please rebase on latest main to get tests fixed

File: install/mysql/glpi-empty.sql [hunk 1]
```diff
@@ -10062,16 +10064,16 @@ DROP TABLE IF EXISTS `glpi_assets_customfielddefinitions`;
...
   `type` varchar(255) NOT NULL,
   `field_options` json,
   `itemtype` VARCHAR(255) NULL DEFAULT NULL,
   `default_value` text,
   `translations` JSON NOT NULL,
   PRIMARY KEY (`id`),
-  UNIQUE KEY `unicity` (`assets_assetdefinitions_id`, `name`),
-  KEY `name` (`name`)
+  UNIQUE KEY `unicity` (`assets_assetdefinitions_id`, `system_name`),
+  KEY `system_name` (`system_name`)
```
Comment (cedric-anne) [inline install/mysql/glpi-empty.sql:10076]: It should be a unique key. People may try to add definitions using SQL queries.

```suggestion
  UNIQUE KEY `system_name` (`system_name`)
```

File: src/Glpi/Asset/CustomFieldDefinition.php [hunk 2]
```diff
@@ -158,10 +158,10 @@ private function validateSystemName(array &$input): bool
         global $DB;
 
         // Spaces are replaced with underscores and the name is made lowercase. Only lowercase letters and underscores are kept.
-        $input['name'] = preg_replace('/[^a-z_]/', '', strtolower(str_replace(' ', '_', $input['name'])));
+        $input['system_name'] = preg_replace('/[^a-z_]/', '', strtolower(str_replace(' ', '_', $input['system_name'])));
         // The name cannot start with an underscore
-        $input['name'] = ltrim($input['name'], '_');
-        if ($input['name'] === '') {
+        $input['system_name'] = ltrim($input['system_name'], '_');
```
Comment (cedric-anne) [inline src/Glpi/Asset/CustomFieldDefinition.php:163]: IMHO, as we are already doing these replacement on client side, we should probably just validate the pattern here, the same way it is done in the `AbstractDefinition::prepareInput()` method.

People who will try to script the creation of custom fields will probably prefer to have an error than an uncontrolled replacement.

File: templates/pages/admin/customobjects/main.html.twig [hunk 3]
```diff
@@ -43,37 +43,27 @@
...
+        {{ __('The value must contains only letters and must not end with "Model" or "Type".') }}
+        {{ __('Also, the value must not be one of the reserved names.') }}
+    {% endset %}
+    {{ fields.textField('label', item.fields['label'], __('Label')) }}
+    {{ fields.textField(
+        'system_name',
+        item.fields['system_name'],
+        __('System name'),
+        {
+            readonly: not item.isNewItem(),
```
Comment (cedric-anne) [inline templates/pages/admin/customobjects/main.html.twig:57]: No need to send the value to the server.

```suggestion
            disabled: not item.isNewItem(),
```

File: templates/pages/admin/customobjects/main.html.twig [hunk 4]
```diff
@@ -137,6 +127,36 @@
...
+                return;
+            }
+            const reserved_names = {{ reserved_system_names|json_encode()|raw }}.map((n) => n.toLowerCase());
+            const existing_names = {{ existing_system_names|json_encode()|raw }}.map((n) => n.toLowerCase());
+            // label is made lowercase and spaces are replaced with underscores. All other characters are removed
+            $('#mainformtable input[name="system_name"]')
+                .val($('#mainformtable input[name="label"]').val().normalize('NFD').toLowerCase().replace(/ /g, '_').replace(/[^a-z_]/g, ''));
+            const system_name = $('#mainformtable input[name="system_name"]').val();
+            if (reserved_names.includes(system_name) || system_name.endsWith('type') || system_name.endsWith('model')) {
+                $('#mainformtable input[name="system_name"]').get(0).setCustomValidity('{{ __('The system name is a reserved name. Please enter a different label or manually change the system name.') }}');
```
Comment (cedric-anne) [inline templates/pages/admin/customobjects/main.html.twig:149]:
```suggestion
                $('#mainformtable input[name="system_name"]').get(0).setCustomValidity('{{ __('The system name is a reserved name. Please enter a different label or manually change the system name.')|e('js') }}');
```

File: templates/pages/admin/customobjects/main.html.twig [hunk 5]
```diff
@@ -137,6 +127,36 @@
...
+            } else if (existing_names.includes(system_name)) {
+                $('#mainformtable input[name="system_name"]').get(0).setCustomValidity('{{ __('The system name is already in use. Please enter a different label or manually change the system name.') }}');
```
Comment (cedric-anne) [inline templates/pages/admin/customobjects/main.html.twig:151]:
```suggestion
                $('#mainformtable input[name="system_name"]').get(0).setCustomValidity('{{ __('The system name is already in use. Please enter a different label or manually change the system name.')|e('js') }}');
```

File: templates/pages/admin/customobjects/main.html.twig [hunk 6]
```diff
@@ -137,6 +127,36 @@
...
+        });
+        function autoUpdateNameField() {
+            if ({{ not item.isNewItem() ? 'true' : 'false' }} || $('#mainformtable input[name="system_name"]').data('manually_changed')) {
+                return;
+            }
+            const reserved_names = {{ reserved_system_names|json_encode()|raw }}.map((n) => n.toLowerCase());
+            const existing_names = {{ existing_system_names|json_encode()|raw }}.map((n) => n.toLowerCase());
+            // label is made lowercase and spaces are replaced with underscores. All other characters are removed
+            $('#mainformtable input[name="system_name"]')
+                .val($('#mainformtable input[name="label"]').val().normalize('NFD').toLowerCase().replace(/ /g, '_').replace(/[^a-z_]/g, ''));
```
Comment (cedric-anne) [inline templates/pages/admin/customobjects/main.html.twig:146]: I think we should let the ability to use uppercase chars here. For instance, our assets class names contains uppercase chars and we will probably want to transform them into generic assets in the future without having to change their system name case.

```suggestion
                .val($('#mainformtable input[name="label"]').val().normalize('NFD').replace(/ /g, '_').replace(/[^a-z_]/g, ''));
```

File: install/mysql/glpi-empty.sql (hunk 1 above)
Comment (cconard96) [reply install/mysql/glpi-empty.sql:10076]: This is already handled by the key above it, no? Custom field system names are only required to be unique for the asset definition they are for.

**Review by cedric-anne — COMMENTED** [review]
Not tested, but it seems technically OK.

**Review by orthagh — CHANGES_REQUESTED** [review]
Seems good so far.
I just have a couple of things to report:

- Majuscules are not reported from label to system name. I think we expect to lower them in the transformation
- The helper for the system name is a bit cryptic. I think it is mostly for a previous iteration, but a refresh could be done, my take (feel free to correct me, or propose better wording):
> The system name field corresponds to what will be used when development is involved. Examples: API calls, webhooks, etc. You can personalize it, but some words are reserved, mostly classes from GLPI, ex Computer, Monitor, etc

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**